File: Objective-C/Reference App/Support/InputOutputViewController.m
//
//  InputOutputViewController.m
//  Reference App
//
//  Created by David Deller on 5/3/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

#import "InputOutputViewController.h"

@interface InputOutputViewController ()

@property (nonatomic, weak) IBOutlet UITextField *inputField;

@end

@implementation InputOutputViewController

- (IBAction)goSubmit:(id)sender {
    [self clearOutput];
    
    if (self.inputField.text != nil) {
        [self handleInput:self.inputField.text];
    }
}

#pragma mark - UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    [self clearOutput];
    
    return YES;
}

- (BOOL)textFieldShouldClear:(UITextField *)textField {
    [self clearOutput];
    
    return YES;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    [self goSubmit:textField];
    
    return NO;
}

#pragma mark -

- (void)handleInput:(NSString *)input {
    // Subclass should override
}

@end

File: Swift 4/Reference App/Support/InputOutputViewController.swift
//
//  InputOutputViewController.swift
//  Reference App
//
//  Created by David Deller on 5/1/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

import UIKit
import TallyGoKit

class InputOutputViewController: OutputViewController, UITextFieldDelegate {

    @IBOutlet weak var inputField: UITextField!
    
    // MARK: -
    
    @IBAction func goSubmit(_ sender: Any?) {
        clearOutput()
        
        if let input = inputField.text {
            handleInput(input)
        }
    }
    
    // MARK: - UITextFieldDelegate
    
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        clearOutput()
        return true
    }
    
    func textFieldShouldClear(_ textField: UITextField) -> Bool {
        clearOutput()
        return true
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        goSubmit(textField)
        return false
    }
    
    // MARK: -
    
    func handleInput(_ input: String) {
        // Subclass should override
    }

}
